infrastructure/src/main/java/org/corfudb/infrastructure/log/BoundedMap.java

```diff
@@ -0,0 +1,116 @@
+package org.corfudb.infrastructure.log;
+
+import com.google.common.base.Preconditions;
+
+import java.util.Iterator;
+import java.util.concurrent.atomic.AtomicLongArray;
+
+public class BoundedMap {
+
+    private final long offset;
+    private final AtomicLongArray array;
+    public static final long NOT_SET = -1;
+
+    public BoundedMap(long offset, int size) {
+        Preconditions.checkArgument(offset >= 0);
+        Preconditions.checkArgument(size > 0);
+        Preconditions.checkArgument(Math.addExact(offset, size) < Integer.MAX_VALUE);
```

Comment:
offset is packed in 36bits, so offset + size can be greater that max integer.

Reply:
That's why I have addExact, it won't overflow. It will fail, if it happens.

That condition won't happen. For example, it will overflow when offset is greater than (2^31– 1 - 10000), at 10k txn/sec that will take (2^31– 1 - 10000) / 10,000 ~ 68 years. Obviously the write rate is way lower than that AND the cluster lifetime is smaller than 68 years.